grpcreflect.go

```diff
@@ -376,3 +407,184 @@ type staticNames struct {
 func (n *staticNames) Names() []string {
 	return n.names
 }
+
+// hackedResolver provides hacks to workaround the fact that connect-grpcreflect-go
+// and connect-grpchealth-go use hacked "connectext.grpc..." packages in the linked-in
+// descriptors (to avoid init-time panics due to conflicts, in case calling code also
+// links in the gRPC runtime's versions of these services).
+type hackedResolver struct {
+	resolver atomic.Pointer[protodesc.Resolver]
+	init     sync.Once
+}
+
+func resolverHackForConnectext() *hackedResolver {
+	res := &hackedResolver{}
+	var delegate protodesc.Resolver = protoregistry.GlobalFiles
+	res.resolver.Store(&delegate)
+	return res
+}
+
+func (r *hackedResolver) FindFileByPath(path string) (protoreflect.FileDescriptor, error) {
+	res := *r.resolver.Load()
+	file, err := res.FindFileByPath(path)
+	if err != nil {
+		if res == protoregistry.GlobalFiles && contains(mangledFileNames, path) {
+			// we need to create "unmangled" versions of the services
+			// that connect libs have to mangle
+			res = r.doInit(res)
+			file, err = res.FindFileByPath(path)
+			if err == nil {
+				return file, nil
+			}
+		}
+		return nil, err
+	}
+	return file, nil
+}
+
+func (r *hackedResolver) FindDescriptorByName(name protoreflect.FullName) (protoreflect.Descriptor, error) {
+	res := *r.resolver.Load()
+	desc, err := res.FindDescriptorByName(name)
+	if err != nil {
+		if res == protoregistry.GlobalFiles && contains(mangledServiceNames, string(name)) {
+			// we need to create "unmangled" versions of the services
+			// that connect libs have to mangle
+			res = r.doInit(res)
+			desc, err = res.FindDescriptorByName(name)
+			if err == nil {
+				return desc, nil
+			}
+		}
+		return nil, err
+	}
+	return desc, nil
+}
+
+func (r *hackedResolver) doInit(original protodesc.Resolver) protodesc.Resolver {
+	res := original
+	r.init.Do(func() {
+		overrides := &protoregistry.Files{}
+		for _, name := range []protoreflect.FullName{ReflectV1ServiceName, ReflectV1AlphaServiceName, healthV1ServiceName} {
+			_, err := original.FindDescriptorByName(name)
+			if err == nil {
+				// this one is fine
+				continue
+			}
+			tryUnmangle(name, overrides)
+		}
+		res = &combinedResolver{first: overrides, second: original}
+		r.resolver.Store(&res)
+	})
+	return res
+}
+
+type combinedResolver struct {
+	first, second protodesc.Resolver
+}
+
+func (r *combinedResolver) FindFileByPath(s string) (protoreflect.FileDescriptor, error) {
+	file, err := r.first.FindFileByPath(s)
+	if err != nil {
+		file, err = r.second.FindFileByPath(s)
+	}
+	return file, err
+}
+
+func (r *combinedResolver) FindDescriptorByName(name protoreflect.FullName) (protoreflect.Descriptor, error) {
+	desc, err := r.first.FindDescriptorByName(name)
+	if err != nil {
+		desc, err = r.second.FindDescriptorByName(name)
+	}
+	return desc, err
```

Comment:
Question: both here an in `FindFileByPath`, we're exposing the error from `.second` if both lookups fail. Is that desirable, or should we instead return the error from `.first`?

Reply:
I can tweak this. We should probably return the error from `.first` unless it is `protoregistry.NotFound`, and then we return the error from `.second` (just in case it has more useful information than simply "not found").

Reply:
Adjusted to only "leak" the error from `second` when the error from first is `NotFound`. This is also what allows the error from initializing the embedded descriptors over in #49 to be observed by a caller. Resolved in 0df528356d1aca3c1b57b7cb9da1b4d2f907016e.